src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -150,8 +155,32 @@ public void resetCursors(final String subscriptionId, final List<NakadiCursor> c
         final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
         validateCursorsBelongToSubscription(subscription, cursors);
         for (final NakadiCursor cursor : cursors) {
-            cursor.validate();
+            cursor.checkStorageAvailability();
         }
+
+        final Map<TopicRepository, List<NakadiCursor>> topicRepositories = cursors.stream().collect(
+                Collectors.groupingBy(
+                        c -> timelineService.getTopicRepository(c.getTimeline())));
+
+        final NakadiCursorComparator comparator = new NakadiCursorComparator(eventTypeCache);
+
+        for (final Map.Entry<TopicRepository, List<NakadiCursor>> entry : topicRepositories.entrySet()) {
+            final List<PartitionStatistics> stats = entry.getKey().loadTopicStatistics(
+                    entry.getValue().stream().map(NakadiCursor::getTimeline).collect(Collectors.toSet()));
+
+            final Map<EventTypePartition, PartitionStatistics> statsMapped = stats.stream()
+                    .collect(Collectors.toMap(
+                            v -> v.getFirst().getEventTypePartition(),
+                            v -> v
+                    ));
+            for (final NakadiCursor toCheck : entry.getValue()) {
+                final PartitionStatistics etpStats = statsMapped.get(toCheck.getEventTypePartition());
+                if (!comparator.isOrdered(etpStats.getBeforeFirst(), toCheck, etpStats.getLast())) {
+                    throw new InvalidCursorException(CursorError.UNAVAILABLE, toCheck);
+                }
+            }
+        }
```

Comment:
I'm not really sure why you created this additional code to check cursors availability. I think we already have code in `TopicRepository.validateReadCursors` that does it (it is currently used to check initial_cursors of newly created subscription)